parse.go

```diff
@@ -84,6 +85,109 @@ func (s *Spec) UnmarshalYAML(node *yaml.Node) error {
 	return nil
 }
 
+func (e *Expect) UnmarshalYAML(node *yaml.Node) error {
+	if node.Kind != yaml.MappingNode {
+		return errors.ExpectedMapAt(node)
+	}
+	// maps/structs are stored in a top-level Node.Content field which is a
+	// concatenated slice of Node pointers in pairs of key/values.
+	for i := 0; i < len(node.Content); i += 2 {
+		keyNode := node.Content[i]
+		if keyNode.Kind != yaml.ScalarNode {
+			return errors.ExpectedScalarAt(keyNode)
+		}
+		key := keyNode.Value
+		valNode := node.Content[i+1]
+		switch key {
+		case "error":
+			if valNode.Kind != yaml.ScalarNode {
+				return errors.ExpectedScalarAt(valNode)
+			}
+			var v string
+			if err := valNode.Decode(&v); err != nil {
+				return err
+			}
+			e.Error = v
+		case "len":
+			if valNode.Kind != yaml.ScalarNode {
+				return errors.ExpectedScalarAt(valNode)
+			}
+			var v *int
+			if err := valNode.Decode(&v); err != nil {
+				return err
+			}
+			e.Len = v
+		case "unknown":
+			if valNode.Kind != yaml.ScalarNode {
+				return errors.ExpectedScalarAt(valNode)
+			}
+			var v bool
+			if err := valNode.Decode(&v); err != nil {
+				return err
+			}
+			e.Unknown = v
+		case "notfound":
+			if valNode.Kind != yaml.ScalarNode {
+				return errors.ExpectedScalarAt(valNode)
+			}
+			var v bool
+			if err := valNode.Decode(&v); err != nil {
+				return err
+			}
+			e.NotFound = v
+		case "json":
+			if valNode.Kind != yaml.MappingNode {
+				return errors.ExpectedMapAt(valNode)
+			}
+			var v *gdtjson.Expect
+			if err := valNode.Decode(&v); err != nil {
+				return err
+			}
+			e.JSON = v
+		case "conditions":
+			if valNode.Kind != yaml.MappingNode {
+				return errors.ExpectedMapAt(valNode)
+			}
+			var v map[string]*ConditionMatch
+			if err := valNode.Decode(&v); err != nil {
+				return err
+			}
+			e.Conditions = v
+		case "matches":
+			if valNode.Kind == yaml.MappingNode {
+				var v map[string]interface{}
+				if err := valNode.Decode(&v); err != nil {
+					return err
+				}
+				e.Matches = v
+			} else if valNode.Kind == yaml.ScalarNode {
+				if valNode.Tag == "!!null" {
```

Comment:
note to @a-hilaly this is actually the only way to tell that someone put the following in their YAML:

```yaml
some:
  nested:
    mapfield:
```

where `some.nested.mapfield` has a `nil` *value* (it's actually valid YAML...)

Reply:
Interesting node tag name ...